python_modules/libraries/dagster-sigma/dagster_sigma/resource.py

```diff
@@ -255,6 +283,87 @@ def try_except_http_warn(self, should_catch: bool, msg: str) -> Iterator[None]:
             else:
                 raise
 
+    def _begin_workbook_materialization(self, workbook_id: str, sheet_id: str) -> str:
+        output = self._fetch_json(
+            f"workbooks/{workbook_id}/materializations",
+            method="POST",
+            json={"sheetId": sheet_id},
+        )
+        return output["materializationId"]
+
+    def _fetch_materialization_status(
+        self, workbook_id: str, materialization_id: str
+    ) -> Dict[str, Any]:
+        return self._fetch_json(f"workbooks/{workbook_id}/materializations/{materialization_id}")
+
+    def _run_materializations_for_workbook(
+        self, workbook_id: str, sheet_ids: AbstractSet[str]
+    ) -> None:
+        materialization_id_to_sheet = dict(
+            zip(
+                [
+                    self._begin_workbook_materialization(workbook_id, sheet_id)
+                    for sheet_id in sheet_ids
+                ],
+                sheet_ids,
+            )
+        )
+        remaining_materializations = set(materialization_id_to_sheet.keys())
+
+        successful_sheets = set()
+        failed_sheets = set()
+
+        while remaining_materializations:
+            materialization_statuses = [
+                self._fetch_materialization_status(workbook_id, materialization_id)
+                for materialization_id in remaining_materializations
+            ]
+            for status in materialization_statuses:
+                if status["status"] not in ("pending", "building"):
+                    remaining_materializations.remove(status["materializationId"])
+                    if status["status"] == "ready":
+                        successful_sheets.add(
+                            materialization_id_to_sheet[status["materializationId"]]
+                        )
+                    else:
+                        failed_sheets.add(materialization_id_to_sheet[status["materializationId"]])
+
+            time.sleep(5)
+
+        if failed_sheets:
+            if successful_sheets:
```

Comment:
I feel fairly strongly that a workbook ought to be a single asset (even if can, in some cases, be partly "materialized") from a catalog point of view. When Fivetran or Airbyte syncs many tables from a source to a destination, each of these tables is a discrete object in the database/warehouse which might be used by different downstream assets, and accessed/viewed separately. 

Even though a workbook is made up of separate sheets/elements, they are all accessed at the same time through viewing the workbook, and I would think most users would reason about them as a single object. I also worry that splitting a workbook into separate sheets would quickly pollute the asset graph - it's a lot easier to comprehend the status, dependencies, and metadata of a single centralized workbook asset rather than having to navigate to each element to compile the list of upstream dependencies.

Reply:
Yeah in hindsight that makes more sense.